Feat/context dsl (#763)
ZupIT/beagle-docs: content/pt/resources/context-dsl.md

Quando inicializamos um contexto populamos um mapa com chave/valor na qual o valor é um objeto qualquer `(Any)` e para conseguirmos alterar esses valores utilizamos o método [`SetContext()`]({{< ref path="/api/actions/setcontext" lang="pt" >}}).

```kotlin
context = ContextData("myBoolean", false),
onInit = listOf(
            SetContext(
                contextId = "myBoolean",
                value = true
            )
```

Após uma análise dos problemas supracitados, vemos que não é possível inferir os tipos do contexto, e além disso, sempre é necessário utilizar o `SetContext` para conseguirmos alterar os valores do contexto. 

Com isso, surgiu a necessidade de criarmos uma DSL Context que revolve o problema de tipagem que altera a maneira de se inicializar um contexto para um `data class`, definindo assim com suas propriedades(chave), bem como abstrai a maneira de atualizar os dados do contexto com o `SetContext` e cria extensions dessas propriedades com métodos que alteram seus valores.

{{% alert color="info" %}}
Basicamente existem 3 tipos de contextos que a DSL Context busca resolver: `Local`, `Global` e `Implícito`.
{{% /alert %}}

## **Como funciona a DSL Context?**

> Através de um processador de anotações são geradas extensions das classes de contexto que são anotadas com **@ContextObject**, na qual para cada propriedade são criadas 2 extensions: `expression` e `change`. A extensão expression é utilizada para conseguir resgatar um Bind.Expression contendo o seu valor e change é utilizado para alterar o valor dessa propriedade.
> 

Reviewer: ```suggestion
A DSL Context funciona por meio de um processador de anotações, onde são geradas extensões das classes de contexto que são anotadas com **@ContextObject** e para cada propriedade são criadas 2 extensões: 

- **`expression`**: Usada para resgatar um **Bind.Expression** contendo o seu valor.
- **`change`**: Usada para alterar o valor dessa propriedade.

```

```kotlin
// extensões geradas
public val MyContext.valueExpression: Bind.Expression<String>
  get() = expressionOf("@{$id.value}")

public fun MyContext.changeValue(value: String): SetContext {
  val contextIdSplit = splitContextId(id)
  return SetContext(
     contextId = contextIdSplit.first,
     path = "${if (contextIdSplit.second != null) "${contextIdSplit.second}." else ""}value",
     value = value
  )
}
```

> {{% alert color="info" %}}
> A DSL Context **não** altera o contrato JSON na qual o Beagle já possui
> {{% /alert %}}

Reviewer: ```suggestion
{{% alert color="info" %}}
A DSL Context **não** altera o contrato JSON que o Beagle já tem.
{{% /alert %}}
```

> ## Requisitos

Reviewer: ```suggestion
## **Requisitos**
```

{{% alert color="info" %}}
A DSL Context só está disponível a partir da versão **X** do Beagle
{{% /alert %}}

> ### Utilizando o Contexto Local com a DSL Context
> 
> Usa-se a anotação **@ContextObject** em data classes que você deseja atribuir como contexto e, obrigatoriamente, esses data classes devem implementar uma interface chamada `Context` que possui uma marcação id. Todos os subtipos que o contexto possa ter também deverão ser anotados.

Reviewer: ```suggestion
##  **Utilizando o Contexto Local com a DSL Context**

Utilize a anotação **@ContextObject** em data classes que você deseja atribuir como contexto e, obrigatoriamente, elas devem implementar uma interface chamada **`Context`** que possui uma marcação ID. 
Todos os subtipos que o contexto possui devem ser anotados. Veja abaixo: 
```

```kotlin
@ContextObject
data class MyContext(
    override val id: String,
    val value: String = "",
    val person: Person = Person("")
) : Context
```

> {{% alert color="info" %}}
> Todas as propriedades declaradas nas data classes devem ser var e possuir um valor default
> {{% /alert %}}
> 
> Com isso, foram geradas as extensions `expression` e `change` para as propriedades value e person no arquivo chamado MyContextNormalizer:

Reviewer: ```suggestion
{{% alert color="info" %}}
Todas as propriedades declaradas nas data classes devem ser **var** e possuir um valor default.
{{% /alert %}}

As extensões **`expression`** e **`change`** foram geradas para as propriedades **value** e **person** no arquivo chamado **MyContextNormalizer**, veja abaixo:
```

  return SetContext(
     contextId = contextIdSplit.first,
     path = "${if (contextIdSplit.second != null) "${contextIdSplit.second}." else ""}value",
     value = value
  )
}

public fun MyContext.changeValue(value: Bind.Expression<String>): SetContext {
  val contextIdSplit = splitContextId(id)
  return SetContext(
     contextId = contextIdSplit.first,
     path = "${if (contextIdSplit.second != null) "${contextIdSplit.second}." else ""}value",
     value = value
  )
}

public val MyContext.personExpression: Bind.Expression<Person>
  get() = expressionOf("@{$id.person}")

public fun MyContext.changePerson(person: Person): SetContext {
  val contextIdSplit = splitContextId(id)
  return SetContext(
     contextId = contextIdSplit.first,
     path = "${if (contextIdSplit.second != null) "${contextIdSplit.second}." else ""}person",
     value = person
  )
}

public fun MyContext.changePerson(person: Bind.Expression<Person>): SetContext {
  val contextIdSplit = splitContextId(id)
  return SetContext(
     contextId = contextIdSplit.first,
     path = "${if (contextIdSplit.second != null) "${contextIdSplit.second}." else ""}person",
     value = person
  )
}
```

> Para inicializarmos um Contexto Local, devemos criar uma instância deste `data class` e atriburimos ele a um contexto de algum componente:

Reviewer: ```suggestion
Para você inicializar um Contexto Local, crie uma instância do `data class` e é necessário atribuir um contexto de algum componente:
```

```kotlin
private var myContext = MyContext("myContext",
        person = Person(
            id = "person",
            firstName = "firstName local",
            lastName = "lastName local")
    )
    Container(context = myContext)
```

> Caso deseje acessar a expressão de uma determinada propriedade utilize a extensão expression:

Reviewer: ```suggestion
Se você acessar a expressão de uma determinada propriedade, utilize a extensão **expression**:
```

```kotlin
Text(myContext.valueExpression),
Text(myContext.person.firstNameExpression),
Text(myContext.person.lastNameExpression)
```

Caso deseje alterar o valor de uma das propriedades, utilize a extensão change:

```kotlin
myContext.person.changeFirstName("new first name")
myContext.changeValue("new value")
```

> ### Utilizando o Contexto Global com a DSL Context
> 

Reviewer: ```suggestion
### Utilizando o Contexto Global com a DSL Context

```
```suggestion
### **Utilizando o Contexto Global com a DSL Context**

```

> Para criar um contexto global, devemos criar um `data class` anotado com **@GlobalContext**:

Reviewer: ```suggestion
Para você criar um contexto global, crie um **`data class`** anotado com **@GlobalContext**:
```

```kotlin
@GlobalContext
data class GlobalObject(
    val street: String = "",
    val person: Person = Person("")
)
```

> {{% alert color="info" %}}
> Todas as propriedades declaradas na `data class` para o contexto global devem ser var e possuir um valor default
> {{% /alert %}}

Reviewer: ```suggestion
{{% alert color="info" %}}
Todas as propriedades declaradas nas `data classes` devem ser **var** e possuir um valor default.
{{% /alert %}}
```

> Da mesma maneira que o contexto local, são criadas extesions para cada propriedade do seu contexto global para conseguirmos acessá-las e alterar seus respectivos valores.

Reviewer: ```suggestion
As extensões são criadas para cada propriedade do seu contexto global, assim você consegue acessá-las e depois alterar seus valores.
```

{{% alert color="warning" %}}
Na aplicação só pode existir uma `data class` anotada com @GlobalContext
{{% /alert %}}

> Para conseguirmos utilizar o contexto global, devemos sempre iniciar uma nova instância da `data class` anotada com o @GlobalContext e através das extensões podemos acesssar/alterar os valores deste contexto:

Reviewer: ```suggestion
Para você utilizar o contexto global, inicie uma nova instância da `data class` anotada com o @GlobalContext. Por meio das extensões, você pode acesssar/alterar os valores deste contexto, veja como:
```

```kotlin
var globalObject = GlobalObject()
globalObject.change(
    GlobalObject("Street A",
        Person("", "first name global", "last name global"))),
globalObject.person.change(Person("", "firstName global changed", "lastName global changed"))
```

> ### Utilizando o Contexto Implicito com a DSL Context
> 
> O Contexto implícito é um contexto que através de um `lambda` você recebe um objeto definido como parâmetro para conseguir acessar possíveis valores que ele receba e manipular/atributir a outros contextos. 
> 
> Para se declarar um contexto implícito deve-se anotar a propriedade com **@ImplicitContext**. Com isso será gerada uma funcão com o mesmo nome do Widget, porém, com a primeira letra minuscula. Este método recebe os mesmo parametros definidos no Widget, entretanto as propriedades que possuem a anotação do contexto implícito são alteradas para um lambda na qual recebe como parametro um objeto primitivo ou um data class que herde de context e retorna o mesmo tipo definido no Widget.

Reviewer: ```suggestion
### **Utilizando o Contexto Implicito com a DSL Context**

O Contexto implícito através de um **`lambda`** recebe um objeto definido como parâmetro, isso é para você acessar os possíveis valores que ele recebe e manipular/atributir a outros contextos. 

Para você declarar um contexto implícito: 
1. Anote a propriedade com **@ImplicitContext**, uma função com o mesmo nome do Widget será gerada, porém, com a primeira letra minúscula. 
Este método recebe os mesmo parâmetros definidos no Widget, porém as propriedades que possuem a anotação do contexto implícito são alteradas para um **lambda** que recebe como parâmetro um objeto primitivo ou um **data class** que herda de um context e retorna o mesmo tipo definido no Widget.
Veja abaixo:
```

```kotlin
@RegisterWidget
class Input(
    val hint: Bind<String>,
    @ImplicitContext
    val onTextChange: List<Action>? = null
) : Widget()

//gerado
public fun input(hint: Bind<String>, onTextChange: ((java.lang.String) -> List<Action>)? = null):
    Input = Input(hint, onTextChange?.invoke(java.lang.String("onTextChange")))
```

> {{% alert color="warning" %}}
> O Contexto implícito somente pode ser usado em custom Widgets que possuem a anotação @RegisterWidget
> {{% /alert %}}
> 
> Agora, para usar esse Widget com o contexto implicito, você deve chamar essa função gerada:

Reviewer: ```suggestion
{{% alert color="warning" %}}
O Contexto implícito pode ser usado somente em **custom Widgets** que possuem a anotação **@RegisterWidget**.
{{% /alert %}}

Se você quiser usar esse Widget com o **contexto implícito**, você deve chamar a função abaixo:
```
